medusa/storage/google_storage.py

```diff
@@ -13,106 +13,292 @@
+                    object_name=object_key,
+                    file_data=src_file,
+                    force_resumable_upload=True,
+                    timeout=-1,
+                )
+        mo = ManifestObject(resp['name'], int(resp['size']), resp['md5Hash'])
+        return mo
+
+    def upload_object_via_stream(self, data: io.BytesIO, container, object_name: str, headers) -> AbstractBlob:
+        loop = self._get_or_create_event_loop()
+        o = loop.run_until_complete(self._upload_object(data, object_name))
+        return o
+
+    async def _upload_object(self, data: io.BytesIO, object_key: str) -> AbstractBlob:
+        logging.debug(
+            '[Storage] Uploading object from stream -> gcs://{}/{}'.format(
+                self.config.bucket_name, object_key
+            )
         )
+        resp = await self.gcs_storage.upload(
+            bucket=self.bucket_name,
+            object_name=object_key,
+            file_data=data,
+            force_resumable_upload=True,
+            timeout=-1,
+        )
+        return AbstractBlob(resp['name'], int(resp['size']), resp['md5Hash'], resp['timeCreated'])
 
-    def download_blobs(self, srcs, dest):
+    def delete_object(self, obj: AbstractBlob):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._delete_object(obj))
 
-        # src is a list of strings, each string is a path inside the backup bucket pointing to a file, or a GCS URI
-        # dst is a table full path to a temporary folder, ends with table name
+    def delete_objects(self, objects: t.List[AbstractBlob]):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._delete_objects(objects))
 
-        # ensure srcs is always a list
-        if isinstance(srcs, str) or isinstance(srcs, Path):
-            srcs = [srcs]
+    async def _delete_objects(self, objects: t.List[AbstractBlob]):
+        coros = [self._delete_object(obj) for obj in objects]
+        await asyncio.gather(*coros)
 
-        generators = self._download_blobs(srcs, dest)
-        return list(itertools.chain(*generators))
+    async def _delete_object(self, obj: AbstractBlob):
+        await self.gcs_storage.delete(
+            bucket=self.bucket_name,
+            object_name=obj.name,
+            timeout=-1,
+        )
 
-    def _download_blobs(self, srcs, dest):
-        with GSUtil(self.config) as gsutil:
-            for parent, src_paths in _group_by_parent(srcs):
-                yield self._download_paths(gsutil, parent, src_paths, dest)
+    def download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._download_blobs(srcs, dest))
+
+    async def _download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        coros = [self._download_blob(src, dest) for src in map(str, srcs)]
+        await asyncio.gather(*coros)
```

Comment:
**Question:** is it downloading all files concurrently at the same time?

Reply:
Yes, the same as with the deletes. Do we chunk by concucurrent transfers?